### PR TITLE
fix(providers/common/sql): add dummy connection setter for backward compatibility

### LINKED ISSUE #42452: MySqlHook constructor fails with exception after upgrading apache-airflow-providers-common-sql from 1.16.0 to 1.17.0
### Apache Airflow Provider(s)

common-sql, mysql

### Versions of Apache Airflow Providers

apache-airflow-providers-common-sql      1.17.0
apache-airflow-providers-mysql           5.6.1

### Apache Airflow version

2.9.2+astro.1

### Operating System

debian astronomer/astro-runtime:11.5.0

### Deployment

Astronomer

### Deployment details

This is an Astronomer-hosted deployment. Docker image is quay.io/astronomer/astro-runtime:11.5.0

### What happened

DAGs that interact with MySqlHook started failing after apache-airflow-providers-common-sql was automatically upgraded from 1.16.0 to 1.17.0.

Here is an example error from the log:

```
[2024-09-24, 18:24:30 UTC] {taskinstance.py:2905} ERROR - Task failed with exception
Traceback (most recent call last):
  File "/usr/local/lib/python3.11/site-packages/airflow/models/taskinstance.py", line 465, in _execute_task
    result = _execute_callable(context=context, **execute_callable_kwargs)
             ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
  File "/usr/local/lib/python3.11/site-packages/airflow/models/taskinstance.py", line 432, in _execute_callable
    return execute_callable(context=context, **execute_callable_kwargs)
           ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
  File "/usr/local/lib/python3.11/site-packages/airflow/models/baseoperator.py", line 401, in wrapper
    return func(self, *args, **kwargs)
           ^^^^^^^^^^^^^^^^^^^^^^^^^^^
  File "/usr/local/lib/python3.11/site-packages/airflow/operators/python.py", line 235, in execute
    return_value = self.execute_callable()
                   ^^^^^^^^^^^^^^^^^^^^^^^
  File "/usr/local/lib/python3.11/site-packages/airflow/operators/python.py", line 252, in execute_callable
    return self.python_callable(*self.op_args, **self.op_kwargs)
           ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
  File "/usr/local/airflow/dags/sftp_handler.py", line 25, in getDBConnection
    return MySqlHook(connection_name).get_conn()
           ^^^^^^^^^^^^^^^^^^^^^^^^^^
  File "/usr/local/lib/python3.11/site-packages/airflow/providers/mysql/hooks/mysql.py", line 75, in __init__
    self.connection = kwargs.pop("connection", None)
    ^^^^^^^^^^^^^^^
AttributeError: property 'connection' of 'MySqlHook' object has no setter
```

requirements.txt had apache-airflow-providers-mysql with no version constraint.

```
apache-airflow-providers-mysql
```

Forcing downgrade to the previous version fixes the problem and allows the DAGs to run again:

```
apache-airflow-providers-mysql
apache-airflow-providers-common-sql <1.17
```

### What you think should happen instead

Because we are including apache-airflow-providers-mysql without a version constraint, I would hope and expect that it would only bring in a version of apache-airflow-providers-common-sql that is compatible with the version of apache-airflow-providers-mysql that is being used. That does not appear to be happening here.

### How to reproduce

1. Install these exact dependency versions:

apache-airflow-providers-common-sql      1.17.0
apache-airflow-providers-mysql           5.6.1

2. Inside a python operator, attempt to get an instance of a MysqlHook with a named connection:

```
conn = MySqlHook("my_named_mysql_connection").get_conn()
```

This will throw an AttributeError:

```
AttributeError: property 'connection' of 'MySqlHook' object has no setter
```

### Anything else

This is consistently reproducible with the listed provider versions.

### Are you willing to submit PR?

- [ ] Yes I am willing to submit a PR!

### Code of Conduct

- [X] I agree to follow this project's [Code of Conduct](https://github.com/apache/airflow/blob/main/CODE_OF_CONDUCT.md)

### PR DESCRIPTION
**Pull request by Lee-W**
## Why

The introduction of [connection property](https://github.com/apache/airflow/pull/40751/files#diff-6e1b2f961cb951d05d66d2d814ef5f6d8f8bf8f43c40fb5d40e27a031fed8dd7R208) breaks `apache-airflow-providers-mysql<5.7.1`, `apache-airflow-providers-elasticsearch<5.5.1` and `apache-airflow-providers-postgres<5.13.0`

## What
Add a dummy connection setter 

Closes: #42452

---
**^ Add meaningful description above**
Read the **[Pull Request Guidelines](https://github.com/apache/airflow/blob/main/contributing-docs/05_pull_requests.rst#pull-request-guidelines)** for more information.
In case of fundamental code changes, an Airflow Improvement Proposal ([AIP](https://cwiki.apache.org/confluence/display/AIRFLOW/Airflow+Improvement+Proposals)) is needed.
In case of a new dependency, check compliance with the [ASF 3rd Party License Policy](https://www.apache.org/legal/resolved.html#category-x).
In case of backwards incompatible changes please leave a note in a newsfragment file, named `{pr_number}.significant.rst` or `{issue_number}.significant.rst`, in [newsfragments](https://github.com/apache/airflow/tree/main/newsfragments).

### CODE REVIEW

File: airflow/providers/common/sql/hooks/sql.py [hunk 1]
```diff
@@ -210,6 +210,17 @@ def connection(self) -> Connection:
             self._connection = self.get_connection(self.get_conn_id())
         return self._connection
 
+    @connection.setter
+    def connection(self, value: Any) -> None:
+        # This dummpy setter is for backward compatibility and should not be used.
```
Comment (phanikumv) [inline airflow/providers/common/sql/hooks/sql.py]:
```suggestion
        # This dummy setter is for backward compatibility and should not be used.
```

File: tests/providers/mysql/hooks/test_mysql.py [hunk 2]
```diff
@@ -67,6 +67,21 @@ def test_get_conn(self, mock_connect):
         assert kwargs["host"] == "host"
         assert kwargs["db"] == "schema"
 
+    @mock.patch("MySQLdb.connect")
+    def test_dummpy_connection_setter(self, mock_connect):
```
Comment (phanikumv) [inline tests/providers/mysql/hooks/test_mysql.py]:
```suggestion
    def test_dummy_connection_setter(self, mock_connect):
```

File: airflow/providers/common/sql/hooks/sql.py [hunk 3]
```diff
@@ -210,6 +210,17 @@ def connection(self) -> Connection:
...
+        # Since the introduction of connection property, the providers listed below
+        # breaks due to assigning value to self.connection
+        #
+        # apache-airflow-providers-mysql<5.7.1
+        # apache-airflow-providers-elasticsearch<5.5.1
+        # apache-airflow-providers-postgres<5.13.0
+        pass
```
Comment (uranusjr) [inline airflow/providers/common/sql/hooks/sql.py:222]: So assigning to this property now silently does nothing… how does this work for those providers?

Comment (Lee-W) [reply airflow/providers/common/sql/hooks/sql.py:222]: they still have `self.connection = ...` which no longer have functionality but cause issue in the older version

Comment (sunank200) [reply airflow/providers/common/sql/hooks/sql.py:222]: +1

**Review by ashb — COMMENTED** [review]
This stops the error, but why does it work? Were those providers never reading `self.connection` again?

**Comment by Lee-W** [comment]
> This stops the error, but why does it work? Were those providers never reading `self.connection` again?

They're reading through the newly introduced property `connection`

**Review by ephraimbuddy — COMMENTED** [review]
Can you patch this change on old common/sql and test with the affected mysql?

My understanding is that the connection would be read from 'get_conn' which validates this change but yeah, let's test with a patch

**Comment by dabla** [comment]
Does the dummy setter solve this issue now.  The problem is caused by my [PR 40751](https://github.com/apache/airflow/pull/40751).

**Comment by Lee-W** [comment]
> Can you patch this change on old common/sql and test with the affected mysql?
> 
> My understanding is that the connection would be read from 'get_conn' which validates this change but yeah, let's test with a patch

Yep, works as expected

**Comment by dabla** [comment]
> > Can you patch this change on old common/sql and test with the affected mysql?
> > My understanding is that the connection would be read from 'get_conn' which validates this change but yeah, let's test with a patch
> 
> Yep, works as expected

Good to hear, sorry for the caused inconvenience.

**Review by phanikumv — APPROVED** [review]

**Review by ephraimbuddy — APPROVED** [review]

**Pull request merged**